### PR TITLE
[Feature][Connector-V2] Support multi-table read for RocketMQ source

### PR DESCRIPTION
**Pull request by kuleat**
### Purpose of this pull request
#10425 
Add multi-table support for RocketMQ source.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add UT and e2e tests.

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [x] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If necessary, please update `incompatible-changes.md` to describe the incompatibility caused by this PR.
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  2. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
  3. Add ci label in [label-scope-conf](https://github.com/apache/seatunnel/blob/dev/.github/workflows/labeler/label-scope-conf.yml)
  4. Add e2e testcase in [seatunnel-e2e](https://github.com/apache/seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/)
  5. Update connector [plugin_config](https://github.com/apache/seatunnel/blob/dev/config/plugin_config)

### CODE REVIEW

**Comment by DanielCarter-stack** [comment]
### Issue 1: Incomplete Null Handling

**Location**: `seatunnel-connectors-v2/connector-rocketmq/src/main/java/org/apache/seatunnel/connectors/seatunnel/rocketmq/source/RocketMqTableIdDeserializationSchema.java:deserialize()`

**Related Context**:
- Parent interface: `org.apache.seatunnel.api.serialization.DeserializationSchema`
- Caller: `RocketMqSourceReader.pollNext()`

**Issue Description**:
When `topicTableMapping` does not contain the current topic and `this.tableId` is also null, the `tableId` variable is null. Subsequent calls to `seaTunnelRow.setTableIdentifier(null)` may cause exceptions in downstream processing.

**Potential Risks**:
- Risk 1: Downstream Sink processing fails due to missing table information
- Risk 2: Data is incorrectly routed or lost in multi-table routing scenarios

**Impact Scope**:
- Direct impact: `RocketMqSourceReader` and downstream Sinks
- Indirect impact: User jobs using multi-table configuration with configuration errors
- Scope: RocketMQ Connector (single Connector)

**Severity**: MAJOR

**Improvement Suggestion**:
```java
TableIdentifier tableId = topicTableMapping.get(message.getTopic());
if (tableId == null) {
    tableId = this.tableId;
    if (tableId == null) {
        throw new SeaTunnelRuntimeException(
            SourceErrorCode.ROCKETMQ_TOPIC_MAPPING_NOT_FOUND,
            String.format("Topic '%s' not found in topic_table_mapping and no default table configured", 
                         message.getTopic())
        );
    }
}
```

**Rationale**: Fail-fast is safer than silent errors, while providing clear error messages to help users debug configuration.

---

### Issue 2: Configuration Naming Inconsistency Risk

**Location**: `seatunnel-connectors-v2/connector-rocketmq/src/main/java/org/apache/seatunnel/connectors/seatunnel/rocketmq/source/RocketMqSourceConfig.java:TOPIC_TABLE_MAPPING`

**Related Context**:
- Configuration system: `org.apache.seatunnel.api.configuration.ConfigOptions`
- Similar implementation: Kafka Connector's `topic` configuration

**Issue Description**:
The code uses `topic_table_mapping` (underscore naming), but the behavior of SeaTunnel's configuration parser needs to be confirmed. If the parser does not support automatic conversion, users must use the underscore form, which is inconsistent with Java-style camelCase naming.

**Potential Risks**:
- Risk 1: Users attempt to use `topicTableMapping` causing configuration parsing failure
- Risk 2: Documentation examples do not match actual code behavior

**Impact Scope**:
- Direct impact: User configuration experience
- Indirect impact: Documentation and code consistency
- Scope: RocketMQ Connector users

**Severity**: MINOR

**Improvement Suggestion**:
1. Confirm SeaTunnel configuration parser's naming conversion rules
2. Explicitly document the required configuration key format in documentation
3. Consider adding configuration alias support

**Rationale**: Configuration system consistency is critical for user experience.

---

### Issue 3: Memory Leak Risk

**Location**: `seatunnel-connectors-v2/connector-rocketmq/src/main/java/org/apache/seatunnel/connectors/seatunnel/rocketmq/source/RocketMqTableIdDeserializationSchema.java:构造函数`

**Related Context**:
- Holder: `RocketMqSourceReader`
- Lifecycle: Entire SourceReader instance lifecycle (during job execution)

**Issue Description**:
`topicTableMapping` is stored as a member variable with no explicit cleanup mechanism. In ultra-large-scale scenarios (e.g., 10000+ topic mappings), these mappings will occupy memory long-term. Even when certain topics are no longer used, the mappings are not cleaned up.

**Potential Risks**:
- Risk 1: Memory usage continues to grow for long-running jobs
- Risk 2: Invalid mappings accumulate in memory during dynamic topic addition/deletion scenarios

**Impact Scope**:
- Direct impact: `RocketMqTableIdDeserializationSchema` instance
- Indirect impact: Overall job memory footprint
- Scope: Large-scale topic mapping scenarios

**Severity**: MAJOR (but limited impact in typical scenarios)

**Improvement Suggestion**:
```java
// Consider using weak references or limiting map size
private final Map<String, TableIdentifier> topicTableMapping;

// Or explicitly suggest in the configuration documentation:
// "For scenarios with more than 1000 topics, consider using multiple jobs instead"
```

**Rationale**: While the issue is not prominent in current scenarios, proactive measures can prevent future production problems.

---

### Issue 4: Concurrency Safety Issue

**Location**: `seatunnel-connectors-v2/connector-rocketmq/src/main/java/org/apache/seatunnel/connectors/seatunnel/rocketmq/source/RocketMqSourceSplitEnumerator.java:addSplits()`

**Related Context**:
- Parent interface: `org.apache.seatunnel.api.source.SourceSplitEnumerator`
- Thread model: Checkpoint thread + event callback threads

**Issue Description**:
Updates to the `snapshotState` field do not use `volatile` or explicit synchronization. Although a new object is created each time (immutable object pattern), in multi-threaded environments, other threads may see stale values.

**Potential Risks**:
- Risk 1: Checkpoint thread reads inconsistent split state
- Risk 2: Under CPU reordering, field reference may update before object content

**Impact Scope**:
- Direct impact: `RocketMqSourceSplitEnumerator` state consistency
- Indirect impact: Checkpoint accuracy and fault recovery
- Scope: RocketMQ Connector's Exactly-once semantics

**Severity**: MAJOR

**Improvement Suggestion**:
```java
private volatile SnapshotState snapshotState;

// Or use AtomicReference
private final AtomicReference<SnapshotState> snapshotState;
```

**Rationale**: On critical paths of distributed systems, thread safety cannot rely on "usually won't cause problems" assumptions.

---

### Issue 5: Insufficient Error Messages

**Location**: `seatunnel-connectors-v2/connector-rocketmq/src/main/java/org/apache/seatunnel/connectors/seatunnel/rocketmq/source/RocketMqTableIdDeserializationSchema.java:deserialize()`

**Related Context**:
- Call chain: RocketMQ Consumer → SourceReader → DeserializationSchema
- Logging framework: SLF4J (SeaTunnel standard)

**Issue Description**:
When a topic is not in the mapping, the code silently falls back to the default `tableId` without logging. This makes it difficult for users to discover configuration errors (e.g., misspelled topic names).

**Potential Risks**:
- Risk 1: User configuration errors remain undetected for long periods
- Risk 2: Data is routed to the wrong table
- Risk 3: Difficult to troubleshoot production environment issues

**Impact Scope**:
- Direct impact: Observability and debugging experience
- Indirect impact: Data correctness
- Scope: All users using multi-table configuration

**Severity**: MAJOR

**Improvement Suggestion**:
```java
TableIdentifier tableId = topicTableMapping.get(message.getTopic());
if (tableId == null) {
    if (LOG.isWarnEnabled()) {
        LOG.warn("Topic '{}' not found in topic_table_mapping, using default table: {}", 
                message.getTopic(), this.tableId);
    }
    tableId = this.tableId;
}
```

**Rationale**: Observability is an important attribute of production systems, and appropriate warning logs can help users quickly locate problems.

---

### Issue 6: Exception Propagation Chain Break

**Location**: `seatunnel-connectors-v2/connector-rocketmq/src/main/java/org/apache/seatunnel/connectors/seatunnel/rocketmq/source/RocketMqSourceReader.java:getTableId()`

**Related Context**:
- Caller: Multiple places within `RocketMqSourceReader`
- Parent class contract: `SeaTunnelRow.getTableIdentifier()` may be null

**Issue Description**:
The `getTableId()` method directly returns `element.getTableIdentifier()`. If `element` is null or `TableIdentifier` is not set, it returns null. The caller may not perform null checks.

**Potential Risks**:
- Risk 1: NPE triggered in caller code
- Risk 2: Error stack trace difficult to trace to root cause

**Impact Scope**:
- Direct impact: `RocketMqSourceReader` internal logic
- Indirect impact: Multi-table routing correctness
- Scope: RocketMQ Connector

**Severity**: MAJOR

**Improvement Suggestion**:
```java
private TableIdentifier getTableId(SeaTunnelRow element) {
    if (element == null) {
        throw new IllegalArgumentException("SeaTunnelRow element cannot be null");
    }
    TableIdentifier tableId = element.getTableIdentifier();
    if (tableId == null) {
        throw new SeaTunnelRuntimeException(
            SourceErrorCode.ROCKETMQ_TABLE_ID_MISSING,
            "SeaTunnelRow does not contain TableIdentifier, multi-table routing failed"
        );
    }
    return tableId;
}
```

**Rationale**: Defensive programming + fail-fast, exposing problems at the source.

---

### Issue 7: Magic Values

**Location**: `seatunnel-connectors-v2/connector-rocketmq/src/main/java/org/apache/seatunnel/connectors/seatunnel/rocketmq/source/RocketMqSourceConfig.java:TOPIC_NAME 和 TABLE_NAME`

**Related Context**:
- Configuration parsing: `TopicTableConfig` inner class
- Configuration file format: User-provided YAML/JSON

**Issue Description**:
Constants `TOPIC_NAME = "topic"` and `TABLE_NAME = "table"` are defined inside the class. If the configuration file format needs to change, modifications must be synchronized across multiple places.

**Potential Risks**:
- Risk 1: Missing modifications in some places during configuration format evolution
- Risk 2: Inconsistent with other Connectors' configuration formats

**Impact Scope**:
- Direct impact: Configuration parsing logic
- Indirect impact: Future configuration compatibility
- Scope: RocketMQ Connector

**Severity**: MINOR

**Improvement Suggestion**:
```java
// Create a separate configuration constants class
public final class RocketMqSourceConfigConstants {
    public static final String TOPIC_NAME = "topic";
    public static final String TABLE_NAME = "table";
    
    private RocketMqSourceConfigConstants() {}
}
```

**Rationale**: Centralized management of configuration constants helps future maintenance and refactoring.

---

### Issue 8: Missing Boundary Testing

**Location**: `seatunnel-connectors-v2/connector-rocketmq/src/test/java/org/apache/seatunnel/connectors/seatunnel/rocketmq/source/RocketMqSourceConfigTest.java` and `RocketMqSourceSplitEnumeratorTest.java`

**Related Context**:
- Testing framework: JUnit 5
- Test coverage: Core configuration and enum logic

**Issue Description**:
Test cases are missing the following boundary conditions:
- Behavior when `topicTableMapping` is an empty Map
- Cases where `topicTableMapping` contains null keys or null values
- Performance testing for large numbers of topic mappings (e.g., 10000+)
- Default table fallback logic when topic does not exist in mapping

**Potential Risks**:
- Risk 1: Undefined behavior when boundary conditions are triggered in production
- Risk 2: Performance issues exposed only after deployment

**Impact Scope**:
- Direct impact: Code quality assurance
- Indirect impact: Production environment stability
- Scope: RocketMQ Connector

**Severity**: MAJOR

**Improvement Suggestion**:
```java
@Test
@DisplayName("Test empty topicTableMapping with default table")
void testEmptyMappingWithDefaultTable() {
    // Test the case of using default table when map is empty
}

@Test
@DisplayName("Test topic not found in mapping")
void testTopicNotFound() {
    // Test behavior when topic is not in the map
    // Verify whether an exception is thrown or a default value is used
}

@Test
@DisplayName("Test large scale mapping performance")
void testLargeScaleMapping() {
    // Test performance with 10000+ topic mappings
    // Verify deserialization performance is within acceptable range
}
```

**Rationale**: Boundary testing is key to ensuring code robustness, especially for production-grade distributed systems.

---

### Issue 9: Incomplete Documentation Examples

**Location**: `docs/en/connectors/source/RocketMQ.md` and `docs/zh/connectors/source/RocketMQ.md`

**Related Context**:
- Documentation standard: Apache SeaTunnel documentation standards
- User group: Data engineers and operations personnel

**Issue Description**:
The documentation added multi-table configuration examples, but may be missing:
- Error scenario descriptions (behavior when topic is not in mapping)
- Priority rules for multi-table vs single-table configuration
- Performance impact notes (recommendations for large numbers of topic mappings)
- Troubleshooting guide

**Potential Risks**:
- Risk 1: Users don't know how to troubleshoot after configuration errors
- Risk 2: Unaware of performance limitations leading to production issues

**Impact Scope**:
- Direct impact: User experience
- Indirect impact: Support cost
- Scope: All RocketMQ Connector users

**Severity**: MINOR

**Improvement Suggestion**:
Add the following sections to the documentation:
1. **Configuration Priority**: Explain priority of `topic_table_mapping` and `table`
2. **Error Handling**: Describe behavior when topic is not in mapping
3. **Performance Considerations**: Recommend not exceeding X topics per job
4. **Troubleshooting**: Common errors and solutions

**Rationale**: Complete documentation is an important component of open source project user experience.

---

### Issue 10: Missing Metrics

**Location**: `seatunnel-connectors-v2/connector-rocketmq/src/main/java/org/apache/seatunnel/connectors/seatunnel/rocketmq/source/RocketMqSourceReader.java`

**Related Context**:
- Metrics system: SeaTunnel Source Metrics specification
- Standard metrics: `RecordsOut`, `BytesOut`, `PendingSplits`

**Issue Description**:
After adding multi-table functionality, the following key metrics are missing:
- Message processing count per topic
- Table routing hit rate (mapping lookup success/failure ratio)
- Deserialization failure statistics (by table or topic dimension)

**Potential Risks**:
- Risk 1: Unable to monitor traffic distribution of multi-table routing in production
- Risk 2: Lack of data support for troubleshooting

**Impact Scope**:
- Direct impact: Observability
- Indirect impact: Operations efficiency
- Scope: All jobs using multi-table functionality

**Severity**: MAJOR

**Improvement Suggestion**:
```java
// Add custom metrics in SourceReader
private final Counter routingHitCounter;
private final Counter routingMissCounter;

// Update metrics during deserialization
if (topicTableMapping.containsKey(message.getTopic())) {
    routingHitCounter.inc();
} else {
    routingMissCounter.inc();
}
```

**Rationale**: In multi-table scenarios, traffic distribution and routing hit rates are key operational metrics.

---

### Issue 11: Deserializer Has Too Many Responsibilities

**Location**: `seatunnel-connectors-v2/connector-rocketmq/src/main/java/org/apache/seatunnel/connectors/seatunnel/rocketmq/source/RocketMqTableIdDeserializationSchema.java`

**Related Context**:
- Design pattern: Decorator pattern
- Future requirements: Content-based dynamic routing

**Issue Description**:
`RocketMqTableIdDeserializationSchema` is responsible for both deserialization and table ID injection. Although the current design is reasonable, if future needs require content-based dynamic routing (e.g., routing based on `type` field in messages), the current design would require significant changes.

**Potential Risks**:
- Risk 1: Future extensions require refactoring core classes
- Risk 2: Mixed responsibilities make code difficult to maintain

**Impact Scope**:
- Direct impact: Architecture extensibility
- Indirect impact: Future feature development
- Scope: RocketMQ Connector

**Severity**: MINOR (reasonable in current scenario, but pay attention to future evolution)

**Improvement Suggestion**:
Consider introducing `RoutingStrategy` interface:
```java
public interface RoutingStrategy {
    TableIdentifier resolveRoute(MessageExt message, TableIdentifier defaultTable);
}

public class TopicNameRoutingStrategy implements RoutingStrategy {
    private final Map<String, TableIdentifier> topicTableMapping;
    
    @Override
    public TableIdentifier resolveRoute(MessageExt message, TableIdentifier defaultTable) {
        return topicTableMapping.getOrDefault(message.getTopic(), defaultTable);
    }
}

// Deserializer is only responsible for injecting routing results
public class RocketMqTableIdDeserializationSchema {
    private final RoutingStrategy routingStrategy;
    
    TableIdentifier tableId = routingStrategy.resolveRoute(message, this.tableId);
}
```

**Rationale**: Strategy pattern can reserve extension points for future routing requirements (content-based, time-window-based, etc.).

---

### Issue 12: Configuration Format Extensibility Limitations

**Location**: `seatunnel-connectors-v2/connector-rocketmq/src/main/java/org/apache/seatunnel/connectors/seatunnel/rocketmq/source/RocketMqSourceConfig.java:TOPIC_TABLE_MAPPING`

**Related Context**:
- Configuration type: `LinkedHashMap<String, TableIdentifier>`
- Limitation: Only supports exact matching

**Issue Description**:
The current configuration format only supports exact topic-to-table mapping. If future needs require:
- Regex-based wildcard matching (e.g., `order-.*` → `order_table`)
- Prefix/suffix-based matching (e.g., `order-prod-*` → `order_table_prod`)

The current design would require significant modifications to configuration types and parsing logic.

**Potential Risks**:
- Risk 1: Future requirements lead to incompatible configuration format changes
- Risk 2: Users need to migrate existing configurations

**Impact Scope**:
- Direct impact: Configuration extensibility
- Indirect impact: Backward compatibility
- Scope: RocketMQ Connector users

**Severity**: MINOR (YAGNI principle, but noteworthy)

**Improvement Suggestion**:
If this need is confirmed, consider supporting multiple configuration formats:
```yaml
# Exact match (current implementation)
topic_table_mapping:
  order-topic: order_table
  payment-topic: payment_table

# Or pattern matching for future extension (example)
topic_table_patterns:
  - pattern: "order-.*"
    table: "order_table"
  - pattern: "payment-(prod|staging)"
    table: "payment_table"
```

**Rationale**: While over-design should be avoided, the configuration system needs to consider future evolution paths.

---

### Issue 13: Type Conversion Safety

**Location**: `seatunnel-connectors-v2/connector-rocketmq/src/main/java/org/apache/seatunnel/connectors/seatunnel/rocketmq/source/RocketMqTableIdDeserializationSchema.java:deserialize()`

**Related Context**:
- Parent interface: `DeserializationSchema<SeaTunnelRow>`
- Call chain: `deserializationSchema.deserialize()` → `setTableIdentifier()`

**Issue Description**:
The code directly modifies the `SeaTunnelRow` object returned by the deserializer:
```java
SeaTunnelRow seaTunnelRow = deserializationSchema.deserialize(message);
seaTunnelRow.setTableIdentifier(tableId);
```

If the underlying deserializer returns an immutable object, shared instance, or cached instance, direct modification will cause:
- Concurrency safety issues (multiple threads sharing the same instance)
- Data contamination (affecting subsequent messages)

**Potential Risks**:
- Risk 1: Data races in multi-threaded environments
- Risk 2: Table ID confusion (one message's table ID overwritten by another)

**Impact Scope**:
- Direct impact: Data correctness
- Indirect impact: Multi-table routing accuracy
- Scope: All jobs using multi-table functionality

**Severity**: CRITICAL

**Improvement Suggestion**:
```java
SeaTunnelRow originalRow = deserializationSchema.deserialize(message);
// Create a new SeaTunnelRow to avoid modifying the original object
SeaTunnelRow seaTunnelRow = new SeaTunnelRow(
    originalRow.getFields(),
    originalRow.getRowKind()
);
seaTunnelRow.setTableIdentifier(tableId);
seaTunnelRow.setMetaData(originalRow.getMetaData());
```

**Rationale**: Defensive programming, not relying on undefined behavioral contracts. Need to confirm whether the `DeserializationSchema` interface allows modification of returned objects.

---

File: seatunnel-connectors-v2/connector-rocketmq/src/main/java/org/apache/seatunnel/connectors/seatunnel/rocketmq/source/RocketMqTableIdDeserializationSchema.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Override
+    public void deserialize(byte[] message, Collector<SeaTunnelRow> out) throws IOException {
+        delegate.deserialize(
+                message,
+                new Collector<SeaTunnelRow>() {
+                    @Override
+                    public void collect(SeaTunnelRow record) {
+                        record.setTableId(tableId);
+                        out.collect(record);
```
[the remark this answers is not in the document]
Comment (kuleat) [reply seatunnel-connectors-v2/connector-rocketmq/src/main/java/org/apache/seatunnel/connectors/seatunnel/rocketmq/source/RocketMqTableIdDeserializationSchema.java:58]: The default `DeserializationSchema.deserialize(byte[], Collector)` already ensures `record != null` before calling `collect()`.

**Comment by kuleat** [comment]
@davidzollo @corgy-w @zhangshenghang Hi, PTAL when you have time. Thanks!

**Review by davidzollo — APPROVED** [review]
+1
Good job

**Review by corgy-w — APPROVED** [review]
